### ISSUE TITLE
General progression?

### ISSUE THREAD

**Issue by yilinwei** [fommil/fp-scala-mortals#1]
I found that my general progression of FP was done in stages:

1. Thinking about state and referential transparency - this made my code easier to reason about and reuse.
2. Higher order functions - this reduced the amount of boilerplate that my code had.
3. More expressive types such as `Reader`, `State` - this meant that the HOF would give me more information and be easier to work with.
4. Understanding typeclasses and laws - this made it easier to reason about the behaviour of FP libraries I was pulling in.
5. abstracting over types - this made code more generic and easier to reuse.
6. abstracting over kinds (such as MTL) - this meant that I had to write less boilerplate

This was over a period of around 1y and each one of the stages was done as a reaction to a particular problem which was encountered at the time.

@zainab-ali for thoughts

**Comment by zainab-ali** [comment]
I started using Scala several months after Yilin.  Coming from a pure OO background,  I was somewhat skeptical with adopting FP.  I kept a progress log, summarized below:

Month 1:
 - Attempt at creating a functional stream, as we didn't know that fs2 existed.  I got stuck and couldn't help with this whatsoever.
- Went over FP in Scala, but failed to recognise use cases for data types other than `List`, `Option` and `Either`.  Could understand typeclasses, but didn't know how to use them.  Recursion was challenging.
- Overused shapeless.  This just made me confused.  There were some terrifying type signatures.
- Also used macros.  Quasiquotes were a lot to stomach. 

Month 2:
- Transitioned to Akka.  Coding was a breeze.  It took us a less than a week to get the skeleton of our system, since we could use familiar OO patterns.

Month 3:
- The code was becoming highly verbose.  Decided to try our hand at functional Scala again.
- Used basic data types in cats such as `Reader` and `State` successfully
- Created  Monad stacks, understood the use of transformers
- This made the code much more concise and easier to test.  I rarely used a debugger after this point.

Month 4:
- Experimented with Free Monads (this was around the time that Freek came out, and frees were all the rage)
- Understood the use of parametricity.  Now able to recognise how and when to create typeclass instances.
- Understood the idea of typeclass laws.

Month 5:
- Transitioned away from frees and towards MTL

**TL;DR:**
1. Understood the idea of mutable state and referential transparency
2. Became with Scala constructs, such as type parameters, higher kinded types, pattern matching, for comprehension
3. Transitioned an Actor model to a monad stack
4. Understood typeclasses and parametricity
5. Understood MTL

**Comment by fommil** [comment]
Ha! This sounds all too familiar. I've written the intro, if you have 10 minutes could you please give it a skim and let me know what you think? PDFs are generated by the self publishing portal, I can get you access if you want.

**Comment by fommil** [comment]
Book is manuscript/book.org ... the other files are generated from it and used by leanpub.

**Comment by zainab-ali** [comment]
I think it's a practical guide for people who want to get their hands dirty with functional programming.
However, I wouldn't say it's directly aimed at Java developers as opposed to the OO paradigm itself.
That's probably a good thing, since most people reading it would be coming from Java and might not think about OO itself.

If you did want to aim at OO, I'd suggest starting with how the paradigms differ.  I had a certain skepticism when starting FP, mainly because:

**In OO names are very important**  
They indicate the responsibilities that something has.  You'd never name a variable `a`  or `b`.  In FP types are far more important than names, so the names end up being shorter and less descriptive, because they don't need to be (`a` is just a value of type `A`)

**OO deals with the structure of code**  
When looking at a problem, an OO developer will ask, "_How do I split this up into responsibilities?  How do these objects talk to each other?_".  An FP developer will ask, "_What are my inputs?  What is my output?  What state do I need in between?_".  

That's a very different way of thinking, and it's hard to believe that it results in a good code structure.  In practice, most functional code turns into library code, so structure is much less of a problem.

**Comment by fommil** [comment]
+1
